misk-hibernate-testing/src/main/kotlin/misk/jdbc/DatabaseAllocator.kt

```diff
@@ -0,0 +1,213 @@
+    latestVersion = databases
+        .filter { it.startsWith(hint) }
+        .max()
+        ?.split("__")
+        ?.last() ?: latestVersion
+
+    var allocated = false
+    var databaseName = "${hint}__${now}__$latestVersion"
+
+    // Keep trying to create a database until we have found an unused name.
+    while (!allocated) {
+      try {
+        databaseName = "${hint}__${now}__$latestVersion"
+        backend.createDatabase(databaseName)
+        allocated = true
+      } catch (e: PersistenceException) {
+        nextVersion()
+      }
+    }
+
+    currentDatabase = databaseName
+    return databaseName
+  }
+
+  /**
+   * Drops all databases that were created by an allocator which are older than the retention
+   * duration of this allocator.
+   */
+  @Throws(PersistenceException::class)
+  fun dropOldDatabases() {
+    val oldDatabases = existingDatabases()
+        .filter {
+          val nameParts = it.split("__").dropLast(1)
+          val date = nameParts.last().toLong()
+          date < now - retentionMillis
+        }
+    for (database in oldDatabases) {
+      backend.dropDatabase(database)
+    }
+  }
+
+  /** A backend for a [DatabaseAllocator]. */
+  interface Backend {
+    /** Defines the type of the data source. */
+    var type: DataSourceType
+
+    /** Returns a list of *all* databases present in the data source. */
+    fun showDatabases(): List<String>
+
+    /**
+     * Drops the indicated database from the data source.
+     *
+     * Throws [PersistenceException] if the database cannot be dropped (i.e. it does not exist).
+     */
+    @Throws(PersistenceException::class)
+    fun dropDatabase(name: String)
+
+    /**
+     * Attempts to create the indicated database in the data source.
+     *
+     * Throws [PersistenceException] if the database already exists.
+     */
+    @Throws(PersistenceException::class)
+    fun createDatabase(name: String)
+  }
+}
+
+/** Implementation of [DatabaseAllocator.Backend] for SQL databases. */
+@Singleton
+class RealDatabaseBackend @Inject constructor(
+  transacterProvider: Provider<Transacter>
+) : DatabaseAllocator.Backend {
+  val transacter: Transacter = transacterProvider.get()
+
+  override var type: DataSourceType = DataSourceType.VITESS
+
+  private fun showDatabasesQuery(): String {
+    return when (type) {
+      DataSourceType.HSQLDB -> "SHOW SCHEMAS"
+      DataSourceType.MYSQL, DataSourceType.VITESS -> "SHOW DATABASES"
+    }
+  }
+
+  private fun dropDatabaseStatement(name: String): String {
+    return when (type) {
+      DataSourceType.HSQLDB -> "DROP SCHEMA $name"
+      DataSourceType.MYSQL, DataSourceType.VITESS -> "DROP DATABASE $name"
+    }
+  }
+
+  private fun createDatabaseStatement(name: String): String {
+    return when (type) {
+      DataSourceType.HSQLDB -> "CREATE SCHEMA $name"
+      DataSourceType.MYSQL, DataSourceType.VITESS -> "CREATE DATABASE $name"
+    }
+  }
+
+  override fun showDatabases(): List<String> {
```

Comment:
should this return a `Set<String>` to show that there are no duplicates?

Reply:
Updated